presto-main/src/main/java/com/facebook/presto/operator/SpatialIndexBuilderOperator.java

```diff
@@ -0,0 +1,222 @@
+    private final PagesSpatialIndexFactory pagesSpatialIndexFactory;
+
+    private final List<Integer> outputChannels;
+    private final int indexChannel;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final Optional<JoinFilterFunctionFactory> filterFunctionFactory;
+
+    private final PagesIndex index;
+    private ListenableFuture<?> indexNotNeeded;
+
+    private boolean finished;
+
+    public SpatialIndexBuilderOperator(
+            OperatorContext operatorContext,
+            PagesSpatialIndexFactory pagesSpatialIndexFactory,
+            List<Integer> outputChannels,
+            int indexChannel,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory,
+            int expectedPositions,
+            PagesIndex.Factory pagesIndexFactory)
+    {
+        this.operatorContext = operatorContext;
+        this.localUserMemoryContext = operatorContext.localUserMemoryContext();
+        this.spatialRelationshipTest = spatialRelationshipTest;
+        this.filterFunctionFactory = filterFunctionFactory;
+
+        this.index = pagesIndexFactory.newPagesIndex(pagesSpatialIndexFactory.getTypes(), expectedPositions);
+        this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+
+        this.outputChannels = outputChannels;
+        this.indexChannel = indexChannel;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return pagesSpatialIndexFactory.getTypes();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+        checkState(!isFinished(), "Operator is already finished");
+
+        index.addPage(page);
+
+        if (!localUserMemoryContext.trySetBytes((index.getEstimatedSize().toBytes()))) {
+            index.compact();
+            localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
+        }
+
+        operatorContext.recordGeneratedOutput(page.getSizeInBytes(), page.getPositionCount());
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (finished) {
+            return;
+        }
+
+        if (indexNotNeeded == null) {
+            PagesSpatialIndex spatialIndex = index.createPagesSpatialIndex(operatorContext.getSession(), indexChannel, spatialRelationshipTest, filterFunctionFactory, outputChannels);
+            localUserMemoryContext.setBytes(spatialIndex.getEstimatedMemorySizeInBytes());
+            indexNotNeeded = pagesSpatialIndexFactory.lendPagesSpatialIndex(spatialIndex);
+        }
+        else if (indexNotNeeded.isDone()) {
```

Comment:
I think you just want an `if` here.  That way if the  pagesSpatialIndexFactory returns an immediately completed future, you can finish immediately.

Reply:
I think you want to use `indexNotNeeded` as the blocked signal, instead of this style where you get a busy wait loop.  Basically, you want to mark this operator as blocked until the `indexNotNeeded` future is completed.  You can add a callback to do `index.clear(),` the memory update, and setting finished when this future completes.  

Note the callback happens on a different thread so you will have to use thread safe references for those operations.  Alternatively, you could move that code to `isFinished`, and change it to

```java
if (!finished && indexNotNeeded != null && indexNotNeeded.isDone()) {
    index.clear();
    localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
    finished = true;
}
return finished;